### PR TITLE
Adding user's info to the navbar

### LINKED ISSUE #76: Add a sidebar to the base_template
## Expected Behavior

When a user is logged into Hotails, they'll have a sidebar that follows them to every page on Hotails, including their profile picture and their nickname.

## Current Behavior

There's no sidebar at the moment.

## Acceptance Criteria

The sidebar should:
- Include the profile picture of the user
- Include the nickname of the user
- Appear on every page on Hotails

### PR DESCRIPTION
**Pull request by OfirMatasas**
### Desired Outcome

Every user on Hotails will have his profile picture and their nickname on the navbar, following to every page on Hotails.

### Implemented Changes

The base_template has changed, that only when a user is logged into Hotails, they'll see this information.

### Connected Issue/Story

Closes #76

### Definition of Done

- Includes the profile picture of the user
- Includes the nickname of the user
- Appears on every page on Hotails

### Test Coverage

This PR includes new unit and integration tests to go with the code

### A Picture of The New Navbar
Untitled

### CODE REVIEW

File: main/views.py [hunk 1]
```diff
@@ -13,7 +15,24 @@ def index(request):
...
+        username = daycare.name
+        dog_name = None
+
+    context = {
+        'profile_picture_url': profile_picture_url,
+        'username': username,
+        'dog_name': dog_name,
+    }
+
+    return render(request, 'main/homepage.html', context)
```
Comment (ErezCohenn) [inline main/views.py:32]: Why are you returning the first dog-owner or daycare object here? Is not it better that we will check if the user exists, and then call for a function of a homepage by the type of the object?
 If it's a dog owner then `dogwner_homepage` and in the same way for daycare? As we did from the beginning and you can see it [here](#63)

**Review by ErezCohenn — COMMENTED** [review]
Nice job!! looks professional 😄 
just take a note of the comment I left

**Comment by tamirmatok** [comment]
extend navbar
Hi @OfirMatasas  , please update PR information with relevant changes (:

File: main/views.py (hunk 1 above)
Comment (tamirmatok) [reply main/views.py]: Hi @ErezCohenn , please note that first method returns the first object matched by the queryset, or None if there is no matching object. In addition ,  At this point we still don't have implementation for dogwner_homepage or daycare_object.
Furthermore , until now the render function include in the request only User information and not the entire object of Day-Care or Dog-Owner(which we need some fields from them for the navbar), so you can see context_processor.py that will add for every request our desirable fields and with this solution we get much shorther views .

**Comment by ormergi** [comment]
ping

**Review by tamirmatok — APPROVED** [review]

File: main/context_processors.py [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+from dogowner.models import DogOwner
+
+
+def navbar_extras(request):
+    navbar_picture_url = navbar_name = None
+    if request.user.is_authenticated:
+        if DogOwner.objects.filter(user=request.user).exists():
+            dog_owner = DogOwner.objects.filter(user=request.user).first()
+            navbar_picture_url = dog_owner.dog_picture_url
+            navbar_name = dog_owner.first_name + ' ' + dog_owner.last_name
```
Comment (jeniaSakirko) [inline main/context_processors.py:11]: you can use the dogowner str functio https://github.com/redhat-beyond/Hotails/blob/b0145b1e6afc5940b6f6ae5be86ecf10ba0d5144/dogowner/models.py#L28

Comment (OfirMatasas) [reply main/context_processors.py]: Changed it.

File: main/context_processors.py [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+
+        elif DayCare.objects.filter(user=request.user).exists():
+            daycare = DayCare.objects.filter(user=request.user).first()
+            navbar_picture_url = Image.objects.filter(daycare_id=daycare).first().url
```
[the remark this answers is not in the document]
Comment (jeniaSakirko) [reply main/context_processors.py]: I'm concerned about the scenario that we did something wrong in the future so the call to `Image.objects.filter(daycare_id=daycare).first()` will return None and then we will get `AttrebuteError` exception
I agree that at this moment is redundant but maybe just create an issue so we won't forget it, btw the default image sounds like great solution :)

[the remark this answers is not in the document]
Comment (OfirMatasas) [reply main/context_processors.py]: Added a default profile picture for dog owner and daycare, and a method that handles a scenario where they have no profile picture :)

**Review by jeniaSakirko — APPROVED** [review]
LGTM

File: daycare/models.py [hunk 4]
```diff
@@ -50,6 +50,11 @@ def create(email, username, password, name, description, price_per_day, capacity
 
         return new_daycare
 
+    def get_daycare_primary_image(self):
```
Comment (Omeramsc) [inline daycare/models.py:53]: I don't see a test for this

File: dogowner/models.py [hunk 5]
```diff
@@ -56,3 +55,8 @@ def create(email, username, password, dog_name,
         new_dog_owner.user.save()
         new_dog_owner.save()
         return new_dog_owner
+
+    def get_dog_owner_profile_image(self):
```
Comment (Omeramsc) [inline dogowner/models.py:59]: I don't see a test for this

File: main/context_processors.py [hunk 6]
```diff
@@ -0,0 +1,21 @@
+from daycare.models import DayCare
+from dogowner.models import DogOwner
+
+
+def navbar_extras(request):
```
Comment (Omeramsc) [inline main/context_processors.py:5]: I don't see a test for this

File: daycare/test_image.py [hunk 7]
```diff
@@ -3,6 +3,8 @@
 from .models import Image
 from django.core.exceptions import ValidationError
 
+DEFAULT_DAYCARE_PROFILE_URL = "../../static/images/daycare-default-profile-image.jpeg"
+
```
Comment (ErezCohenn) [inline daycare/test_image.py:8]: you can add the `DEFAULT_DAYCARE_PROFILE_URL` to Conftest.py file instead of using a constant multiple times

File: main/tests.py [hunk 8]
```diff
@@ -19,6 +21,11 @@ def create_dog_owner_user():
                            )
 
 
+@pytest.fixture()
+def create_request_factory():
+    return RequestFactory()
+
```
Comment (ErezCohenn) [inline main/tests.py:27]: can you rename the test methods? if someone does not know what is `RequestFactory` so this test is understandable

File: static/CSS/navbar.css [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+#logoutHref {
+    background-color:#1e1e1e;
+    border-radius:12px;
+    color:#E19512;
+    font-weight:bold;
+    float:right;
+    padding: 5px 10px 5px 10px;
+    font-size:1rem;
+    margin:1px 0 0 20px;
+}
```
Comment (ErezCohenn) [inline static/CSS/navbar.css:29]: there are 2 folders for CSS files, can you squash them for one? consider open a new issue about it

**Review by ErezCohenn — CHANGES_REQUESTED** [review]

**Comment by Omeramsc** [comment]
Some parts of this PR are identical to #67 , but there is only 1 commit. 
please fix it by separating the commits of the old and new content.

**Review by ErezCohenn — APPROVED** [review]

**Comment by jeniaSakirko** [comment]
> Some parts of this PR are identical to #67 , but there is only 1 commit. please fix it by separating the commits of the old and new content.

@OfirMatasas ^^

**Comment by OfirMatasas** [comment]
> > Some parts of this PR are identical to #67 , but there is only 1 commit. please fix it by separating the commits of the old and new content.
> 
> @OfirMatasas ^^

After a talk with Omer, the plan is to merge the dog owner homepage before this PR, and then after rebasing the common code will be part of the master, and it will be ok.

**Comment by ErezCohenn** [comment]
> > > Some parts of this PR are identical to #67 , but there is only 1 commit. please fix it by separating the commits of the old and new content.
> > 
> > 
> > @OfirMatasas ^^
> 
> After a talk with Omer, the plan is to merge the dog owner homepage before this PR, and then after rebasing the common code will be part of the master, and it will be ok.

so don't forget to add to the description of the PR that this PR depends on #67  :)

File: dogowner/test_dog_owner.py [hunk 10]
```diff
@@ -163,3 +163,14 @@ def test_dog_owner_user_creation_with_url_not_an_image(self):
                             dog_picture_url="https://www.not/an/image.com",
                             dog_age=10, dog_weight=6, dog_gender='M'
                             )
+
+    def test_dogowner_has_customized_profile_image(self, create_dog_owner_user):
+        dogowner_profile_image = create_dog_owner_user.get_dog_owner_profile_image_url()
+        assert dogowner_profile_image != "../../static/images/dog-owner-default-profile-picture.jpeg"
+        assert dogowner_profile_image is not None
```
Comment (Yuval-Vino) [inline dogowner/test_dog_owner.py:170]: this test is redundant if profile_image equals the wanted image, so of course its not None

File: dogowner/test_dog_owner.py [hunk 11]
```diff
@@ -163,3 +163,14 @@ def test_dog_owner_user_creation_with_url_not_an_image(self):
...
+
+    def test_dogowner_has_default_profile_image(self, create_dog_owner_user):
+        create_dog_owner_user.dog_picture_url = None
+        dogowner_profile_image = create_dog_owner_user.get_dog_owner_profile_image_url()
+        assert dogowner_profile_image == "../../static/images/dog-owner-default-profile-picture.jpeg"
+        assert dogowner_profile_image is not None
```
Comment (Yuval-Vino) [inline dogowner/test_dog_owner.py:176]: same comment as above

File: main/tests.py [hunk 12]
```diff
@@ -77,3 +77,18 @@ def test_dog_owner_homepage_is_visible_for_dog_owner(self, client, create_dog_ow
...
+        client.force_login(user=create_dog_owner_user.user)
+        response = client.get("/homepage/")
+        assert response.context['navbar_picture_url'] == create_dog_owner_user.dog_picture_url
+        assert response.context['navbar_name'] == create_dog_owner_user.__str__()
+
+    def test_navbar_getting_daycare_nickname_and_profile_picture(self, client, create_daycare_user):
+        client.force_login(user=create_daycare_user.user)
+        response = client.get("/homepage/")
+        assert response.context['navbar_picture_url'] == create_daycare_user.get_daycare_primary_image_url()
+        assert response.context['navbar_name'] == create_daycare_user.name
```
Comment (Yuval-Vino) [inline main/tests.py:94]: same comment as above

File: main/context_processors.py [hunk 13]
```diff
@@ -0,0 +1,21 @@
...
+    navbar_picture_url = navbar_name = None
+    if request.user.is_authenticated:
+        if DogOwner.objects.filter(user=request.user).exists():
+            dog_owner = DogOwner.objects.filter(user=request.user).first()
+            navbar_name = dog_owner.__str__()
+            navbar_picture_url = dog_owner.get_dog_owner_profile_image_url()
+        elif DayCare.objects.filter(user=request.user).exists():
```
Comment (Yuval-Vino) [inline main/context_processors.py:12]: maybe you can just use else here (if a user is logged and he is not a dogowner we know he is a daycare), and avoid line 6 of init variables to None

File: daycare/test_image.py [hunk 14]
```diff
@@ -25,9 +24,10 @@ def test_image_creation_with_invalid_image_url(self, create_daycare_user):
...
-        assert daycare_profile_image != DEFAULT_DAYCARE_PROFILE_URL
+        assert daycare_profile_image != "../../static/images/daycare-default-profile-image.jpeg"
         assert daycare_profile_image is not None
 
-    def test_daycare_has_default_profile_image_when_no_customized_picture_was_found(self, create_daycare_user):
+    def test_daycare_has_default_profile_image(self, create_daycare_user):
         daycare_profile_image = create_daycare_user.get_daycare_primary_image_url()
-        assert daycare_profile_image == DEFAULT_DAYCARE_PROFILE_URL
+        assert daycare_profile_image == "../../static/images/daycare-default-profile-image.jpeg"
+        assert daycare_profile_image is not None
```
Comment (Omeramsc) [inline daycare/test_image.py:33]: redundant. the previous assertion already confirms it.

File: dogowner/models.py [hunk 15]
```diff
@@ -56,3 +55,8 @@ def create(email, username, password, dog_name,
         new_dog_owner.user.save()
         new_dog_owner.save()
         return new_dog_owner
+
+    def get_dog_owner_profile_image_url(self):
+        if self.dog_picture_url is not None:
```
[the remark this answers is not in the document]
Comment (Omeramsc) [reply dogowner/models.py:60]: would also make the default profile picture path as a constant.

File: dogowner/test_dog_owner.py [hunk 16]
```diff
@@ -163,3 +163,15 @@ def test_dog_owner_user_creation_with_url_not_an_image(self):
                             dog_picture_url="https://www.not/an/image.com",
                             dog_age=10, dog_weight=6, dog_gender='M'
                             )
+
+    def test_dogowner_has_customized_profile_image(self, create_dog_owner_user):
+        dogowner_profile_image = create_dog_owner_user.get_dog_owner_profile_image_url()
+        dog_owner_default_picture_url = "../../static/images/dog-owner-default-profile-picture.jpeg"
+        assert dogowner_profile_image != dog_owner_default_picture_url
+        assert dogowner_profile_image is not None
```
Comment (Omeramsc) [inline dogowner/test_dog_owner.py:171]: same as I commented about daycare

**Review by Omeramsc — CHANGES_REQUESTED** [review]
Left some comments

File: dogowner/models.py [hunk 17]
```diff
@@ -56,3 +58,6 @@ def create(email, username, password, dog_name,
         new_dog_owner.user.save()
         new_dog_owner.save()
         return new_dog_owner
+
+    def get_dog_owner_profile_image_url(self):
+        return self.dog_picture_url if self.dog_picture_url else DOG_OWNER_DEFAULT_PROFILE_PICTURE_URL
```
Comment (ormergi) [inline dogowner/models.py:63]: Why isnt this logic part of create()?

File: main/tests.py [hunk 18]
```diff
@@ -76,4 +76,25 @@ def test_dog_owner_homepage_is_visible_for_dog_owner(self, client, create_dog_ow
         client.force_login(user=create_dog_owner_user.user)
         response = client.get("/homepage/")
         assert response.status_code == 200
-        assert list(response.context['daycares']) == list(DayCare.objects.all())
+        list_of_daycares_from_response = list(response.context['daycares'])
+        list_of_all_daycares = list(DayCare.objects.all())
+        assert list_of_daycares_from_response == list_of_all_daycares
```
Comment (ormergi) [inline main/tests.py:81]: This is not related to the scope of this PR, right?

File: main/tests.py [hunk 19]
```diff
@@ -76,4 +76,25 @@ def test_dog_owner_homepage_is_visible_for_dog_owner(self, client, create_dog_ow
...
+
+
+@pytest.mark.django_db
+class TestNavbarView:
+    def test_navbar_getting_dog_owner_nickname_and_profile_picture(self, client, create_dog_owner_user):
+        client.force_login(user=create_dog_owner_user.user)
+        response = client.get("/homepage/")
+        dog_owner_picture_url = create_dog_owner_user.dog_picture_url
+        assert response.context['navbar_picture_url'] == dog_owner_picture_url
```
Comment (ormergi) [inline main/tests.py:90]: I would pass variables substitutions in cases like this, and assert the actual value directly instead:
```
assert response.context['navbar_picture_url'] ==  create_dog_owner_user.dog_picture_url
```
Also, it would make the test look nicer if the `assert` line would be separated with new line from the test setup logic.

File: main/context_processors.py (hunk 6 above)
Comment (ormergi) [inline main/context_processors.py:5]: Why is it necessary to filler out the authenticated user from the models?
Is it possible to use the logged in user in this context?

File: conftest.py [hunk 20]
```diff
@@ -8,6 +8,11 @@
 from review.models import Review
 
 
+DOG_OWNER_FIXTURE_PROFILE_PICTURE_URL = "https://www.akc.org/wp-content/uploads/2019/06/Bohemian-Shepherd.1.jpg"
```
Comment (ormergi) [inline conftest.py:11]: Please move this change (and similar changes) to their own commit.

**Review by ormergi — COMMENTED** [review]
Hi, this is nice feature!

There are some formatting and refactoring changes that makes it harder to review this PR,
please separate them to different commits /PRs before the commit that introduce the new logic.

**Review by Omeramsc — APPROVED** [review]
This PR has room to improve, and Or's comments should've been addressed.
however it is good enough, and I won't want this work to not be shown at the last demo, therefore I am merging it. 

If you will continue to work on this project later, please address the comments.
